backend/LexBoxApi/Services/SendReceiveService.cs

```diff
@@ -0,0 +1,39 @@
+﻿using LexCore.Entities;
+using LexCore.ServiceInterfaces;
+using LexCore.Utils;
+using Nito.AsyncEx;
+
+namespace LexBoxApi.Services;
+
+public class SendReceiveService : ISendReceiveService
+{
+    private readonly ConcurrentWeakDictionary<string, AsyncReaderWriterLock> _projectLocks = new();
+    private readonly CancellationToken _cancelled = new(true);
+
+    /// <summary>
+    /// Notify that a project is being sent & received. This will block migration from starting. It will return null if a migration has started indicating that send receive should be blocked
+    /// </summary>
+    /// <returns>null if you shouldn't start, a disposable to dispose when you're done</returns>
+    public async ValueTask<IDisposable?> BeginSendReceive(string projectCode)
+    {
+        var projectLock = _projectLocks.GetOrAdd(projectCode, _ => new AsyncReaderWriterLock());
```

Comment:
I just realized that this will block other S/R's.
That's probably ok 🤷, it's just interesting, because hg presumably handles this itself as well.

Reply:
I'm not sure what you mean. But this is using a read write lock. Which allows many readers, but only one writer and not both. In this context a 'write' is used to block S&R. And an S&R just counts as a read.

Reply:
Oh, you're right of course. I see now.